src/socket/mod.rs

```diff
@@ -98,6 +106,13 @@ impl<'a, 'b> Socket<'a, 'b> {
             where F: FnMut(&IpRepr, &IpPayload) -> Result<R, Error> {
         dispatch_socket!(self, |socket [mut]| socket.dispatch(timestamp, limits, emit))
     }
+
+    pub(crate) fn as_raw(&mut self) -> Option<&mut RawSocket<'a, 'b>> {
+        match *self {
+            Socket::Raw(ref mut raw) => Some(raw),
+            _ => None,
+        }
+    }
```

Comment:
Oh, this is a different issue, `as_socket` doesn't let us interrogate. Sure. Let's add `try_as_socket` in addition to `as_socket` to the `AsSocket` trait, that would be much cleaner.